### PR TITLE
A few changes required to remove 'capture'.

### PR DESCRIPTION
**Pull request by tpoterba**
Better numeric parsing story.
Integer bounds checking.
Use JSON to communicate objects in py4j.
to_expr uses literals if possible.
Literals use broadcast if they're not the scalar primitives.

### CODE REVIEW

**Comment by tpoterba** [comment]
master, sample.vcf:
```
In [2]: %timeit ds.rows().collect()
6.87 s ± 62.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

This branch:
```
In [3]: %timeit ds.rows().collect()
311 ms ± 12.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

File: python/hail/expr/expression.py [hunk 1]
```diff
@@ -681,7 +736,7 @@ def collect(self):
         """
         uid = Env._get_uid()
         t = self._to_table(uid)
-        return t.aggregate(hl.agg.collect(t[uid]))
+        return [r[uid] for r in t.collect()]
```
Comment (cseed) [inline python/hail/expr/expression.py]: Note to self: I had an example from the tutorial where this change made a HUGE difference.  Check again after this lit stuff settles down.

File: python/hail/expr/functions.py [hunk 2]
```diff
@@ -38,6 +38,50 @@ def null(t):
...
+
+    Returns
+    -------
+    :class:`.Expression`
+    """
+    if dtype is None:
+        try:
+            dtype = impute_type(x)
+        except ExpressionException:
+            raise
```
Comment (cseed) [inline python/hail/expr/functions.py]: Isn't this the same as no try ... except?

File: python/hail/expr/types.py [hunk 3]
```diff
@@ -154,6 +180,9 @@ class TInt32(Type):
     In Python, these are represented as :obj:`int`.
     """
 
+    _min_value = -(1 << 31)
+    _max_value = (1 << 31) - 1
```
Comment (cseed) [inline python/hail/expr/types.py]: Not sure why these are here, can just return in min_value, etc?

File: python/hail/expr/types.py [hunk 4]
```diff
@@ -269,6 +320,8 @@ def __str__(self):
     def _eq(self, other):
         return isinstance(other, TFloat64)
 
+    def _convert_to_json(self, x):
+        return float(x)
```
Comment (cseed) [inline python/hail/expr/types.py]: By typecheck, it already has to be float, doesn't it?

Comment (tpoterba) [reply python/hail/expr/types.py]: meant to be convert_from_json. JSON returns "NaN" strings

Comment (tpoterba) [reply python/hail/expr/types.py]: it'll fail the tutorial without this, I think.

Comment (tpoterba) [reply python/hail/expr/types.py]: yep, failed exactly where I expected. I fixed without retesting locally.

File: python/hail/expr/types.py (hunk 3 above)
Comment (tpoterba) [reply python/hail/expr/types.py]: yeah, true.

File: python/hail/expr/expression.py (hunk 1 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: we can actually optimize it further by selecting just 'uid' frmo the table. I bet the slow thing is collecting Loci, which go through java again.

File: src/main/scala/is/hail/expr/Parser.scala [hunk 5]
```diff
@@ -409,8 +409,14 @@ object Parser extends JavaTokenParsers {
     }
 
   def primary_expr: Parser[AST] =
-    withPos("""-?\d+(\.\d+)?[eE][+-]?\d+[dD]?""".r) ^^ (r => Const(r.pos, r.x.toDouble, TFloat64())) |
+    withPos("f32#" ~> """-?\d+(\.\d+)?[eE][+-]?\d+""".r) ^^ (r => Const(r.pos, r.x.toFloat, TFloat32())) |
+    withPos("f64#" ~> """-?\d+(\.\d+)?[eE][+-]?\d+""".r) ^^ (r => Const(r.pos, r.x.toDouble, TFloat64())) |
+      withPos("f32#" ~> """-?\d*(\.\d+)?""".r) ^^ (r => Const(r.pos, r.x.toFloat, TFloat32())) |
+      withPos("f64#" ~> """-?\d*(\.\d+)?""".r) ^^ (r => Const(r.pos, r.x.toDouble, TFloat64())) |
+      withPos("""-?\d+(\.\d+)?[eE][+-]?\d+[dD]?""".r) ^^ (r => Const(r.pos, r.x.toDouble, TFloat64())) |
       withPos("""-?\d*\.\d+[dD]?""".r) ^^ (r => Const(r.pos, r.x.toDouble, TFloat64())) |
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply src/main/scala/is/hail/expr/Parser.scala]: it requires lots of changes to existing scala stuff. another day, definitely.

File: python/hail/expr/expression.py (hunk 1 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: I'll do this change here.

File: python/hail/expr/expression.py [hunk 6]
```diff
@@ -76,109 +76,191 @@ def construct_reference(name, type, indices, prefix=None):
...
+        else:
+            raise ValueError("Hail has no integer data type large enough to store {}".format(x))
+    elif isinstance(x, float):
+        return tfloat64
+    elif isinstance(x, str):
+        return tstr
+    elif isinstance(x, Locus):
+        return tlocus(x.reference_genome)
+    elif isinstance(x, Interval):
+        return tlocus(x.reference_genome)
```
Comment (cseed) [inline python/hail/expr/expression.py]: tinterval

File: python/hail/expr/expression.py [hunk 7]
```diff
@@ -76,109 +76,191 @@ def construct_reference(name, type, indices, prefix=None):
...
+    elif x is None:
+        raise ExpressionException("Hail cannot impute the type of 'None'")
+    else:
+        raise ExpressionException("Hail cannot automatically impute type of {}: {}".format(type(x), x))
+
+
+def to_expr(e, dtype=None):
+    if isinstance(e, Expression):
+        if dtype:
+            assert e.dtype == dtype, 'expected {}, got {}'.format(dtype, e.dtype)
```
Comment (cseed) [inline python/hail/expr/expression.py]: This should be a user-friendly message, not an assert.  For example, Table.parallelize(v, t) might error out here if v isn't of type t.

**Pull request merged**